plugin/src/main/groovy/com/btkelly/gnag/tasks/GnagReportTask.java

```diff
@@ -90,18 +95,68 @@ public void setGitHubExtension(GitHubExtension gitHubExtension) {
     }
 
-    private void updatePRStatus(GitHubStatusType gitHubStatusType) {
-
+    private void fetchPRShaIfRequired() {
         if (StringUtils.isBlank(prSha)) {
-            GitHubPullRequest pullRequestDetails = gitHubApi.getPullRequestDetails();
+            GitHubPRDetails pullRequestDetails = gitHubApi.getPRDetailsSync();
 
             if (pullRequestDetails != null && pullRequestDetails.getHead() != null) {
                 prSha = pullRequestDetails.getHead().getSha();
             }
         }
+    }
 
+    private void updatePRStatus(GitHubStatusType gitHubStatusType) {
         if (StringUtils.isNotBlank(prSha)) {
-            gitHubApi.postUpdatedGitHubStatus(gitHubStatusType, prSha);
+            gitHubApi.postUpdatedGitHubStatusAsync(gitHubStatusType, prSha);
+        }
+    }
+
+    private void postViolationComments(@NotNull final Set<Violation> violations) {
+        final Set<Violation> violationsWithAllLocationInfo
+                = ViolationsUtil.getViolationsWithAllLocationInfo(violations);
+
+        if (StringUtils.isBlank(prSha) || violationsWithAllLocationInfo.isEmpty()) {
+            gitHubApi.postGitHubPRCommentAsync(ViolationsFormatter.getHtmlStringForAggregatedComment(violations));
+            return;
+        }
+        
+        final List<Diff> diffs = gitHubApi.getPRDiffsSync();
+        
+        if (diffs.isEmpty()) {
+            gitHubApi.postGitHubPRCommentAsync(ViolationsFormatter.getHtmlStringForAggregatedComment(violations));
+            return;
+        }
+
+        for (final Violation violation : ViolationsUtil.getViolationsWithValidLocationInfo(violations, diffs)) {
+            //noinspection ConstantConditions
+            final int offsetDiffLineNumber = ViolationsUtil.getOffsetDiffLineNumberForViolation(violation, diffs);
+            
+            //noinspection ConstantConditions
+            gitHubApi.postGitHubInlineCommentSync(
+                    ViolationFormatter.getHtmlStringForInlineComment(violation),
+                    prSha,
+                    violation.getRelativeFilePath(),
+                    offsetDiffLineNumber);
+        }
+        
+        final Set<Violation> violationsWithMissingOrInvalidLocationInfo = 
+                ViolationsUtil.getViolationsWithMissingOrInvalidLocationInfo(violations, diffs);
+        
+        if (!violationsWithMissingOrInvalidLocationInfo.isEmpty()) {
+            try {
+                /*
+                 * Try to post the aggregate comment _strictly after_ all individual comments. GitHub seems to round
+                 * post times to the nearest second, so delaying by one whole second should be sufficient here.
+                 */
+                Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+            } catch (final InterruptedException e) {
+                e.printStackTrace();
+            }
+
+            gitHubApi.postGitHubPRCommentAsync(
+                    ViolationsFormatter.getHtmlStringForAggregatedComment(
```

Comment:
Could refactor `ViolationsFormatter.getHtmlStringForAggregatedComment` and `ViolationsUtil.getViolationsWithMissingOrInvalidLocationInfo` into a single method that returns a `Map`, to avoid searching twice.